src/Engine/Segments/SegmentEvaluator.php

```diff
@@ -0,0 +1,187 @@
+<?php
+
+namespace Flagsmith\Engine\Segments;
+
+use Flagsmith\Engine\Environments\EnvironmentModel;
+use Flagsmith\Engine\Identities\IdentityModel;
+use Flagsmith\Engine\Identities\Traits\TraitModel;
+use Flagsmith\Engine\Utils\HashingTrait;
+
+class SegmentEvaluator
+{
+    use HashingTrait;
+    /**
+     * Get the identity segments.
+     * @param EnvironmentModel $environment
+     * @param IdentityModel $identity
+     * @param array|null $overrideTraits
+     * @return array
+     */
+    public static function getIdentitySegments(
+        EnvironmentModel $environment,
+        IdentityModel $identity,
+        array $overrideTraits = null
+    ): array {
+        return array_filter(
+            $environment->getProject()->getSegments()->getArrayCopy(),
+            fn (SegmentModel $segment) => self::evaluateIdentityInSegment(
+                $identity,
+                $segment,
+                $overrideTraits
+            )
+        );
+    }
+
+    /**
+     * evaluate identity in segment.
+     * @param IdentityModel $identity
+     * @param SegmentModel $segment
+     * @param array|null $overrideTraits
+     * @return bool
+     */
+    public static function evaluateIdentityInSegment(
+        IdentityModel $identity,
+        SegmentModel $segment,
+        array $overrideTraits = null
+    ): bool {
+        $rulesCount = count($segment->getRules());
+
+        if (empty($rulesCount)) {
+            return false;
+        }
+
+        $traitSegmentRules = [];
+        foreach ($segment->getRules() as $rule) {
+            $segmentRule = self::traitsMatchSegmentRule(
+                $overrideTraits ?? $identity->getIdentityTraits()->getArrayCopy(),
+                $rule,
+                $segment->getId(),
+                $identity->compositeKey()
+            );
+
+            $traitSegmentRules[] = $segmentRule;
+
+            if (!$segmentRule) {
+                break;
+            }
+        }
+
+        return self::all($traitSegmentRules);
```

Comment:
Since we're doing the evaluation at each stage of the loop above, do we need to use this `all` function? Couldn't we just return `true` here and then above return `false` if any of the rule evaluations fail?